### PR TITLE
Scene ref/unref

### PR DESCRIPTION
**Pull request by ubitux**

### CODE REVIEW

File: libnopegl/src/scene.c [hunk 1]
```diff
@@ -49,11 +54,13 @@ static void detach_root(struct ngl_scene *s)
 
 struct ngl_scene *ngli_scene_dup(struct ngl_scene *s)
 {
-    struct ngl_scene *copy = ngli_memdup(s, sizeof(*s));
+    struct ngli_scene *copy = ngli_memdup(s->internal, sizeof(*copy));
```
Comment (mbouron) [inline libnopegl/src/scene.c]: can you use some `src`/`dst` prefixes over there ? something like dst_{public,private} and starts by storing `s->internal` in local variable so you can use its sizeof.

Comment (ubitux) [reply libnopegl/src/scene.c]: this code goes away later on in the PR; the dup function is replaced by a ref function that looks like this:
```c
struct ngl_scene *ngl_scene_ref(struct ngl_scene *s)
{
    struct ngli_scene *priv = s->internal;
    NGLI_RC_REF(priv);
    return s;
}
```

File: libnopegl/src/nopegl.h.in [hunk 2]
```diff
@@ -175,7 +175,9 @@ NGL_API char *ngl_scene_dot(const struct ngl_scene *scene);
  *
  * @param ss    pointer to the pointer to the nope.gl context
  */
-NGL_API void ngl_scene_freep(struct ngl_scene **sp);
+NGL_API void ngl_scene_unrefp(struct ngl_scene **sp);
```
Comment (mbouron) [inline libnopegl/src/nopegl.h.in:178]: In the commit msg: replace `ngl_scene_freep()` with `ngl_scene_unrefp()`.

File: libnopegl/src/nopegl.h.in [hunk 3]
```diff
@@ -177,8 +177,6 @@ NGL_API char *ngl_scene_dot(const struct ngl_scene *scene);
  */
 NGL_API void ngl_scene_unrefp(struct ngl_scene **sp);
 
-#define ngl_scene_freep ngl_scene_unrefp
```
Comment (mbouron) [inline libnopegl/src/nopegl.h.in]: I would remove the function directly in the previous commit. IHMO, there is no point in keeping the intermediate state and that would simplify the history log.

File: libnopegl/src/nopegl.h.in [hunk 4]
```diff
@@ -104,7 +104,7 @@ struct ngl_node;
  */
 struct ngl_scene {
     void *internal;         /* reserved internal pointer */
-    struct ngl_node *root;  /* root node of the scene graph */
+    struct ngl_node *root;  /* root node of the scene graph, must be set with the ngl_scene_init_from_* functions */
```
Comment (mbouron) [inline libnopegl/src/nopegl.h.in:107]: I would something like `set by ngli_scene_init_from_* functions (must not be set by the user)`. While at it, I would mention that the fields below are user writable.

File: libnopegl/src/nopegl.h.in (hunk 3 above)
Comment (ubitux) [reply libnopegl/src/nopegl.h.in]: done

File: libnopegl/src/nopegl.h.in (hunk 4 above)
Comment (ubitux) [reply libnopegl/src/nopegl.h.in]: Branch reworked to use `ngl_scene_params`, so these commits were not relevant anymore and thus are removed in this last iteration.

The PR is now on top of https://github.com/NopeForge/nope.gl/pull/199

File: CHANGELOG.md [hunk 5]
```diff
@@ -38,6 +40,9 @@ Versioning](https://semver.org/spec/v2.0.0.html) for `libnopegl`.
   textureSize()/imageSize() instead (`%s_dimensions` is still available for 2D
   textures)
 
+### Removed
+- `ngl_scene_freep` function; use `ngl_scene_unrefp` instead
+
```
Comment (mbouron) [inline CHANGELOG.md:45]: `api: add ngl_scene_unrefp() and use it` -> `api: replace ngl_scene_freep() with ngl_scene_unrefp()`

File: libnopegl/src/api.c [hunk 6]
```diff
@@ -361,7 +361,7 @@ int ngli_ctx_configure(struct ngl_ctx *s, const struct ngl_config *config)
         s->scene = old_scene; // restore detached scene on error
         goto fail;
     }
-    ngl_scene_freep(&old_scene); // ngli_ctx_set_scene() copied the scene safely so we drop the old one
+    ngl_scene_unrefp(&old_scene); // ngli_ctx_set_scene() incremented the ref counter of the scene so we safely unref here
```
Comment (mbouron) [inline libnopegl/src/api.c:364]: nit: so we *can* safely unref here ?

File: libnopegl/src/scene.c (hunk 1 above)
Comment (mbouron) [reply libnopegl/src/scene.c]: The commit has been dropped with your new iteration, so resolved :)

**Review by mbouron — APPROVED** [review]
Except from my last comments, LGTM.

File: libnopegl/src/api.c (hunk 6 above)
Comment (ubitux) [reply libnopegl/src/api.c]: yeah, added

File: CHANGELOG.md (hunk 5 above)
Comment (ubitux) [reply CHANGELOG.md:45]: fixed

**Pull request merged**